cluster-autoscaler/core/scaledown/planner/planner.go

```diff
@@ -0,0 +1,271 @@
+	"k8s.io/autoscaler/cluster-autoscaler/processors/status"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/clustersnapshot"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/utilization"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	pod_util "k8s.io/autoscaler/cluster-autoscaler/utils/pod"
+
+	apiv1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	klog "k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+type eligibilityChecker interface {
+	FilterOutUnremovable(context *context.AutoscalingContext, scaleDownCandidates []*apiv1.Node, timestamp time.Time, unremovableNodes *unremovable.Nodes) ([]string, map[string]utilization.Info)
+}
+
+type removalSimulator interface {
+	DropOldHints()
+	CheckNodeRemoval(node string, podDestinations map[string]bool, timestamp time.Time, pdbs []*policyv1.PodDisruptionBudget) (*simulator.NodeToBeRemoved, *simulator.UnremovableNode)
+}
+
+// Planner is responsible for deciding which nodes should be deleted during scale down.
+type Planner struct {
+	context              *context.AutoscalingContext
+	unremovableNodes     *unremovable.Nodes
+	unneededNodes        *unneeded.Nodes
+	rs                   removalSimulator
+	scheduler            *scheduling.HintingSimulator
+	latestUpdate         time.Time
+	eligibilityChecker   eligibilityChecker
+	nodeUtilizationMap   map[string]utilization.Info
+	actuationStatus      scaledown.ActuationStatus
+	resourceLimitsFinder *resource.LimitsFinder
+}
+
+// New creates a new Planner object.
+func New(context *context.AutoscalingContext, processors *processors.AutoscalingProcessors, deleteOptions simulator.NodeDeleteOptions) *Planner {
+	resourceLimitsFinder := resource.NewLimitsFinder(processors.CustomResourcesProcessor)
+	return &Planner{
+		context:              context,
+		unremovableNodes:     unremovable.NewNodes(),
+		unneededNodes:        unneeded.NewNodes(processors.NodeGroupConfigProcessor, resourceLimitsFinder),
+		rs:                   simulator.NewRemovalSimulator(context.ListerRegistry, context.ClusterSnapshot, context.PredicateChecker, simulator.NewUsageTracker(), deleteOptions, true),
+		scheduler:            scheduling.NewHintingSimulator(context.PredicateChecker),
+		eligibilityChecker:   eligibility.NewChecker(processors.NodeGroupConfigProcessor),
+		nodeUtilizationMap:   make(map[string]utilization.Info),
+		resourceLimitsFinder: resourceLimitsFinder,
+	}
+}
+
+// UpdateClusterState needs to be periodically invoked to provide Planner with
+// up-to-date information about the cluster.
+func (p *Planner) UpdateClusterState(podDestinations, scaleDownCandidates []*apiv1.Node, as scaledown.ActuationStatus, pdb []*policyv1.PodDisruptionBudget, currentTime time.Time) errors.AutoscalerError {
+	p.latestUpdate = currentTime
+	p.actuationStatus = as
+	// TODO: clone cluster snapshot to avoid persisting changes done by the
```

Comment:
Oh, I thought the comment meant forking just for injecting the ongoing actuation, not the whole `UpdateClusterState` call. But even then, isn't there something non-obvious that could stop working if we were to drop the cluster snapshot changes between `UpdateClusterState` and `NodesToDelete`? E.g. some limit/PDB calculations?

Also - why would the post-scale-down logic be better off with the cluster state unmodified by scale-down? Is it because it's too "ephemeral" (removed from what the actual scheduler will do)? Just curious, these are non-blocking questions.

Reply:
It is just the scale down simulation can modify the snapshot significantly while being essentially an implementation detail of Planner: binpacking simulations during scale up are kind of similar in that manner, CA doesn't persist binpacked pods in the snapshot either.

I guess if we integrated scheduler and CA more tightly and started actuating cluster snapshot to actually reflect CA simulations (schedule pods or even move them between nodes if live migration ever becomes feasible in k8s), persisting more changes in various parts of CA would be worth considering, but I'm not convinced it makes sense today.

Reply:
You're right, somehow I thought we might persist something in the snapshot between scale-up and scale-down already, but we don't, and it makes sense. Thanks